### DISCUSSION TITLE
Option to invert point at which measurement is read.

### DISCUSSION THREAD

**Discussion by MGM92** [bhowiebkr/laser-level-webcam#31] [Ideas]
It would be great if there was an option to invert the point at which the measurement is taken to be the center of the darkest area; this would enable the program to take readings from a taut wire in front of a light source as well as lasers. Using a wire in front of the sensor gives a very sharp dark peak, which could be used to overcome some of the issues with laser beams not giving a sufficiently distinct point/line for reliable readings. preliminary testing, of this appears very promising relative to the Java program that was originally developed. Unfortunately, I don't have any coding skills to attempt to effect such a solution, though I would be very keen to try it if someone with the necessary programming skills were able to add this functionality to this excellent application.

**Comment by box185** [comment]
A wire placed in front of the sensor does make a very thin shadow across the face of the sensor, and it's easy to modify line 155 of the Workers.py file - changing min_value, max_value = histo.min(), histo.max() to min_value, max_value = histo.max(), histo.min().

The problem with a wire attached in front of the sensor is that the wire and the sensor move together, which means that the shadow does not move relative to the sensor. Moving the wire away from the sensor creates a shadow that is broadly diffused.

**Reply by MGM92** [reply]
Hi Box185,
Thanks for your response and the solution, I havent yet tried it but really appreciate the coding help and am keen to try this. The idea of using a taut wire is that it is not attached but is independent of the sensor, and does not move at all. 
The taut wire method is actually a very established means of measuring flatness in the installation of very large cnc machines would precision straight edges would be impractical (picture the type Boeing might use to machine wing components) and is, in some cases, preferred to very expensive laser systems. As an example, if you wanted to check the flatness of a linear guide rail, you would suspend a wire between two points such that the wire is parrallel to, but independent of, the guide rail. Weights would be used to tension the wire (typically to just shy of its breaking point is my understanding). There was an apparatus similar to a microscope that would then be mounted to a carriage on the rail, and would be situated such that the viewer could look into it an view the wire against a background which would allow detection of changes in the relative height of the wire. In this case, the webcam sensor would replace the aforementioned microscope-like apparatus. My intention was to mount the webcam sensor to the inside of a 'U' shaped structure and a bright light to the other side, such that the light shines on the sensor. This would then be suspended such that the two parts sandwich the wire very close to touching but without the wire making any contact with any part of the apparatus. There are then formulas to calculate the catenary sag, which should allow very accurate measurements. The proceure should be well reported with visual aids because it is a well established procedure used in industry.
So far I have been using a precision bubble level and a heel-to-toe series of measurements which I believe is working well, as the measurements are generally very repeatable, but I would like to corroborate my results with this device (which should also be a lot quicker as it doesn't require minutes with each movement for a bubble to settle.) 
Kind regards,
Mike

**Comment by box185** [comment]
It seems to me that the 'microscope' used in the method you wrote about would be able to focus on the wire at some distance. If that were the case, then this could work. There would be no possibility to focus the bare sensor on a distant wire - which is what I tried to write about earlier. Without a lens on the webcam sensor, the wire at a distance would be completely blurred and useless.

I did not consider using the webcam with the built in lens, and I don't know what the DOF ( depth-of-field ) is for the built in lens. In my test, I placed the wire directly in front of the sensor and this worked really well - only because it was not blurred. It nearly dissappeared as I moved the wire away from the bare sensor.

**Reply by MGM92** [reply]
Hi Box185,
If you're trying to measure something like the flatness of a linear guide rail, it shouldn't be too difficult to align the taught wire to within a couple of mm of parrallel with the rail. If you then have an apparatus holding both the sensor and a light either side of the wire and move it up the length of the rail, you should be able to have the sensor stay within a couple of mm of the wire. I believe this kind of distance still gave a very sharp/crisp shadow on the sensor when I tried it (I did have the light probably a couple of inches away). If you are looking to take measurements directly on a surface, rather than on a rail, you could clamp something like a carpenter's level or a plasterer's feather edge, to act as a straight reference then align the taut wire with this to achieve a similarly consistent distance between the sensor and wire. An alternative, though more complex, would be to rig up a means of holding the sensor that allows fine lateral adjustment to fine tune the distance between the sensor and the wire. 
If you're looking to align two separated surfaces into the same plane, such as paired rail mounting surfaces for a cnc mill, I would set up the taut wire in the middle between the two surfaces, such that the apparatus could simply be turned around and measure from second surface with reference to the opposite side of the wire. This way you could essentially determine a plane spanning quite substantiall distances without the sensor ever needing to be more than a tiny distance from the wire. You are likely to need to determine that the sensor mounting apparatus is levelled in the plane perpendicular to the wire, but this is not likely to be prohibitively difficult with a precision level, which can be picked up very economically.
I hope this is helpful.
Kind regards,
Mike
